src/libmongoc/src/mongoc/mongoc-database.c

```diff
@@ -996,6 +995,293 @@ mongoc_database_create_collection (mongoc_database_t *database,
+   mongoc_collection_t *eccCollection = NULL;
+   mongoc_collection_t *ecocCollection = NULL;
+   mongoc_collection_t *dataCollection = NULL;
+   bool ok = false;
+   bson_t *cc_opts = NULL;
+   
+   /* Create ESC collection. */
+   escName = _mongoc_get_encryptedField_state_collection (
+      encryptedFields, name, "esc", error);
+   if (!escName) {
+      goto fail;
+   }
+
+   escCollection = create_collection (database, escName, NULL /* opts */, error);
+   if (!escCollection) {
+      goto fail;
+   }
+
+   /* Create ECC collection. */
+   eccName = _mongoc_get_encryptedField_state_collection (
+      encryptedFields, name, "ecc", error);
+   if (!eccName) {
+      goto fail;
+   }
+
+   eccCollection = create_collection (database, eccName, NULL /* opts */, error);
+   if (!eccCollection) {
+      goto fail;
+   }
+
+   /* Create ECOC collection. */
+   ecocName = _mongoc_get_encryptedField_state_collection (
+      encryptedFields, name, "ecoc", error);
+   if (!ecocName) {
+      goto fail;
+   }
+
+   ecocCollection = create_collection (database, ecocName, NULL /* opts */, error);
+   if (!ecocCollection) {
+      goto fail;
+   }
+
+   /* Create data collection. */
+   cc_opts = bson_copy (opts);
+   if (!BSON_APPEND_DOCUMENT (cc_opts, "encryptedFields", encryptedFields)) {
+      bson_set_error (error,
+                        MONGOC_ERROR_COMMAND,
+                        MONGOC_ERROR_COMMAND_INVALID_ARG,
+                        "unable to append encryptedFields");
+      goto fail;
+   }
+   dataCollection = create_collection (database, name, cc_opts, error);
+   if (!dataCollection) {
+      goto fail;
+   }
+
+   /* Create index on __safeContent__. */
+   {
+      bson_t *keys = BCON_NEW ("__safeContent__", BCON_INT32 (1));
+      char *index_name;
+      bson_t *create_indexes;
+
+      index_name = mongoc_collection_keys_to_index_string (keys);
+      create_indexes = BCON_NEW ("createIndexes",
+                                 BCON_UTF8 (name),
+                                 "indexes",
+                                 "[",
+                                 "{",
+                                 "key",
+                                 BCON_DOCUMENT (keys),
+                                 "name",
+                                 BCON_UTF8 (index_name),
+                                 "}",
+                                 "]");
+
+      if (!mongoc_database_write_command_with_opts (database,
+                                                    create_indexes,
+                                                    NULL /* opts */,
+                                                    NULL /* reply */,
+                                                    error)) {
+         bson_destroy (create_indexes);
+         bson_free (index_name);
+         bson_destroy (keys);
+         goto fail;
+      }
+      bson_destroy (create_indexes);
+      bson_free (index_name);
+      bson_destroy (keys);
```

Comment:
These cleanups are the same. I would try:

```c
ok = mongoc_database_write_command_with_opts(...);
// [cleanup]
```

No `goto` needed since the return value is effectively stored in `ok`